### ISSUE TITLE
implementing `xmap` + `SerialLoop` behavior for mini-batching using `shmap`

### ISSUE THREAD

**Issue by vadmbertr** [google/jax#20272]
Hi,

As in https://github.com/google/jax/discussions/15807, I would like to use `xmap` and `SerialLoop` to parallelize over multiple devices while "mini-batching" as vectorizing over the whole axis of my array causes OOM errors. The code snippet from https://github.com/google/jax/discussions/15807 can be used (I'm not solving an ODE, but that's not really the point):

```python
import os
os.environ["XLA_FLAGS"] = '--xla_force_host_platform_device_count=8' # Use 8 CPU devices

import numpy as np
import jax
import jax.numpy as jnp
from jax.sharding import Mesh
from jax.experimental.maps import xmap, SerialLoop
from jax.experimental.ode import odeint

def rhs(y, t):
    return t**2

def f(a):
    return odeint(rhs, y0=jnp.array(a), t=jnp.array([0., 1.]))[-1]

input_array = np.random.rand(16)
devices = np.array(jax.devices())

with Mesh(devices, ("x",)):
    out_array = xmap(
        f,
        in_axes=["map_axis", ...],
        out_axes=["map_axis", ...],
        axis_resources={"map_axis": ("x", SerialLoop(2))}
    )(input_array)
```
I'm also facing `unbound axis name` error, and I was hoping for a fix. 

Now that `xmap` has been deprecated, I'm looking at `shmap` (as advised in the docs) but it is unclear to me how to achieve this "mini-batching" without explicitely looping through my batched dimension. Did I missed something?

Thanks,
Vadim

**Comment by superbobry** [comment]
Hi Vadim,

It looks like you need a combination of `shard_map`, which would handle device parallelism, and either `lax.vmap` or manual slicing to implement the mini-batching. Re-using the definitions from your example,

```
from jax.experimental.shard_map import shard_map
P = jax.sharding.PartitionSpec

def serial_f(x, n=2):
  assert x.shape[0] % n == 0
  result = jax.lax.map(f, jnp.reshape(x, (n, -1, *x.shape[1:])))
  return jnp.reshape(result, (-1, *result.shape[2:]))

shard_map(serial_f, mesh, P("x"), P("x"), check_rep=False)(input_array)
```

or

```
def serial_f(x, n=2):
  return jnp.concatenate([f(chunk) for chunk in jnp.split(x, n)])

shard_map(serial_f, mesh, P("x"), P("x"), check_rep=False)(input_array)
```

Lmk if that works for your use-case.

**Comment by vadmbertr** [comment]
Hi,

It's a bit of both, here is the simplified version of my implementon (I did not check if that runs):

```python
from functools import partial

from jax.experimental.shard_map import shard_map
from jax.sharding import PartitionSpec as P

# need to vectorize over leading axis
@partial(jax.vmap, in_axes=(0,), out_axes=(0,))
def vmap_f(x):
    return f(x)

# need to parallelize over leading axis
@partial(shard_map, mesh=mesh, in_specs=(P("x"),), out_specs=(P("x"),), check_rep=False)
def shmap_f(x, n=2):
    return jnp.concatenate([vmap_f(chunk) for chunk in jnp.split(x, n)])

shmap_f(input_array)
```

I'm mainly reporting to know if I missed something, and because we are encouraged to do so by `xmap` deprecation warning. It felt to me that the `SerialLoop` hook was quite handy, but this approach is not overwhelming neither.

**Comment by superbobry** [comment]
Thanks for sharing Vadim! 

Yeah, I agree `SerialLoop` provided a slightly more succinct way of encoding this, but it is only one of the many capabilities of `xmap` which jointly were a little bit overwhelming. This is why we now recommend slightly more verbose (and hopefully more clear) alternatives via `shmap` and friends.

**Comment by superbobry** [comment]
I will close this, since there are no actions necessary on our side.
